13-offer-encoding.md

```diff
@@ -0,0 +1,589 @@
+   than the modern satoshi-based counting.
+6. The bech32 encoding was found to have an issue with extensions,
+   which means we want to replace or discard it anyway.
+7. The `payment_secret` designed to prevent probing by other nodes in
+   the path was only useful if the invoice remained private between the
+   payer and payee.
+8. Invoices must be given per-user, and are actively dangerous if two
+   payment attempts are made for the same user.
+
+
+# Payment Flow Scenarios
+
+Here I use "user" as shorthand for the individual user's lightning
+node, and "merchant" as the shorthand for the node of someone who is
+selling or has sold something.
+
+There are two basic payment flows supported by BOLT 13:
+
+The general user-pays-merchant flow is:
+1. A merchant publishes an *offer*, such as on a web page or a QR code.
+2. Every user requests a unique *invoice* over the lightning network
+   using an *invrequest* message.
+3. The merchant replies with the *invoice*.
+4. The user makes a payment to the merchant indicated by the invoice.
+
+The merchant-pays-user flow (e.g. ATM or refund):
+1. The merchant provides a user-specific *invrequest* in a webpage or QR code,
+   with an amount (for a refund, also a reference to the to-be-refunded
+   invoice).
+2. A user sends an *invoice* for the amount in the *invrequest* (for a
+   refund, a proof that they requested the original)
+3. The merchant makes a payment to the user indicated by the invoice.
+
+## Payment Proofs and Payer Proofs
+
+Note that the normal lightning "proof of payment" can only demonstrate that an
+invoice was paid (by showing the preimage of the `payment_hash`), not who paid
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+1. Sharing the minimum information required to prove sections of the
+   invoice in dispute (e.g. the description of the item, the payment
+   hash, and the merchant's signature).
+2. Contain a transferrable proof that the user is the one who was
+   responsible for paying the invoice in the first place.
+
+Providing a key in *invrequest* allows a user to prove that they were the one
+to request the invoice.  In addition, the merkle construction of the BOLT 13
+invoice signature allows the user to selectively reveal fields of the invoice
+in case of dispute.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the designated prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).
+
+FIXME: Encoding/decoding requirements.
+- Any required padding bits are set to zero / reject if not zero.
+
+## Rationale
+
+The use of bech32 is arbitrary, but already exists in the bitcoin
+world.  We omit the six-character trailing checksum since all uses
+here involve a signature.
```

Comment:
Computational cost for signature verification and checksum is very different. Checksum meant for fast and simple verification in the UI before working with encoded data within the business logic layer of an app. I propose to leave checksum even if we have the signature

Reply:
We do need *some* string representation though.  The others all suck even worse.

We could use a binary format in a QR in *theory*, but in practice I don't think it would work well (readers tend to want things to look like a URL)

Reply:
What about URs? 
https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md

Reply:
What about URs? 
https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md